### PR TITLE
TOOLS-2688: mongodump requires EOF when passing in the password as STDIN

### PR DESCRIPTION
**Pull request by varsha804**
I've added a call to `ioutil.ReadAll()` in `readPassFromStdin()` so that we can check if a password is being piped into a mongo tool command. The subsequent `os.Stdin.Read()` call will have an error string of `EOF` if this is the case.

### CODE REVIEW

**Comment by varsha804** [comment]
There aren't any written tests for password logic, so I tested manually:
Screen Shot 2020-08-18 at 1 42 22 PM

File: password/password.go [hunk 1]
```diff
@@ -50,10 +51,20 @@ func Prompt() (string, error) {
 // we aren't using a terminal for standard input
 func readPassFromStdin() (string, error) {
 	pass := []byte{}
+	var err error
+	fi, _ := os.Stdin.Stat()
+
+	if (fi.Mode() & os.ModeCharDevice) == 0 {
+		log.Logv(log.DebugLow, "reading data from stdin pipe")
```
Comment (varsha804) [inline password/password.go:58]: Not sure if this is the right log level.

Comment (varsha804) [reply password/password.go:58]: Oh, good point. I'll make those adjustments!

Also, I was planning on tagging reviewers once Huan merged his PR! I'll add Tim after I re-request your review.

**Review by tfogo — CHANGES_REQUESTED** [review]
The original reporter of TOOLS-2688 said that the issue was piping a password without an EOF is not handled by the tools. This is incorrect. The error thrown is "EOF". So we _**are**_ receiving an EOF and the problem is actually that we aren't properly handling EOF.

The statement that the original reporter made that they were piping a password "without an EOF" doesn't quite make sense. EOF is not a special character like EOT or ETX that can be part of a string, but a condition of a stream. `echo -n "foo" | mongodump` will result in mongodump seeing an EOF from stdin because echo will close the pipe.

All we need to do to fix this is add this after `os.Stdin.Read(chBuf[:])`:

```go
if err == io.EOF {
    break
}
```

The problem was we were only breaking on newline, carriage return, ETX, and EOT. But we never handled EOF. `eofKey` is actually the code point for EOT and `eotKey` is actually the code point for ETX. There is no such thing as an EOF character. It's an annoying misnomer because it made it look like we were handling EOF, but we were not.

So I think we only need to rename those two constants and add that error check for io.EOF.

**Comment by tfogo** [comment]
Although, I'll add that what you have now works perfectly, except we need to rename eotKey and eofKey. I think the only reason we might not want to use `ioutil.readAll()` is performance. But given that passwords shouldn't be very large, I don't think that should be a huge concern. Any thoughts?

**Comment by varsha804** [comment]
@tfogo I think the only advantage of the current implementation is that it simplifies the for-loop by removing those if statements. Might be better to go with your implementation since it'll be performant when a password is unusually long.

I tried adding a unit test for the case where stdin is a terminal, but `terminal.MakeRaw()` in `pass_util.go` was making that difficult. Unless anyone has any objections, I'll just resort to the usual manual testing for this case.

File: password/pass_util.go [hunk 2]
```diff
@@ -21,10 +21,10 @@ import (
 // operating systems that aren't solaris
 
 func IsTerminal() bool {
-	return terminal.IsTerminal(int(syscall.Stdin))
```
Comment (varsha804) [inline password/pass_util.go:24]: Redundant type conversion.

File: password/pass_util.go [hunk 3]
```diff
@@ -41,5 +41,5 @@ func GetPass() (string, error) {
 	if err != nil {
 		return "", err
 	}
-	return string(pass), nil
```
Comment (varsha804) [inline password/pass_util.go:44]: Redundant type conversion

**Review by tfogo — APPROVED** [review]
I think manual testing is fine for that. LGTM!

**Comment by tfogo** [comment]
I just noticed it's failing on windows though - the type cast of `syscall.Stdin` to int is necessary because `syscall.Stdin` is actually of type `syscall.Handle` on Windows.

**Pull request merged**